### PR TITLE
fix: resolve Matter transaction deadlock on goHome and handle Idle mode (fixes #273)

### LINKED ISSUE #273: sdjqr02rr - Xiaomi roborock mi robot - Pressing RETURN TO DOCK does not work
**What is your RVC model?**
`sdjqr02rr - Xiaomi roborock mi robot`

**Describe the bug**
I am able to start/pause the vacuum correctly using the tile inside home app.

however, after I tap PAUSE (which works), when I tap RETURN TO DOCK, the RVC does nothing. stands still.

**To Reproduce**

Press START CLEANING (play symbol)
Press PAUSE (pause symbol)
Press RETURN TO DOCK

info[20:16:34.881][InteractionServer]Invoke « @1:1a40f361•e867⇵5d27 invokes: MA_roboticvacuumcleaner:0x1.RvcRunMode:0x54.changeToMode:0x0
info[20:16:34.888][ProtocolService]Invoke « AspirapolvereTavernetta.AspirapolvereTavernetta-1706100013688.rvcRunMode.changeToMode @1:1a40f361•e867⇵5d27✉0b4b2e6c newMode: 2
info[20:16:34.890][Aspirapolvere Tavernetta]Changing mode to 2 (endpoint AspirapolvereTavernetta-1706100013688.1)
info[20:16:34.894][Matterbridge Xiaomi Roborock Plugin][Name=Aspirapolvere Tavernetta][Model=rockrobo.vacuum.v1] Initiating full cleaning...
info[20:16:35.931][ProtocolService]Invoke » AspirapolvereTavernetta.AspirapolvereTavernetta-1706100013688.rvcRunMode.changeToMode @1:1a40f361•e867⇵5d27✉0b4b2e6c status: 0 statusText: Running
info[20:16:42.400][InteractionServer]Invoke « @1:1a40f361•e867⇵5d28 invokes: MA_roboticvacuumcleaner:0x1.RvcOperationalState:0x61.pause:0x0
info[20:16:42.403][ProtocolService]Invoke « AspirapolvereTavernetta.AspirapolvereTavernetta-1706100013688.rvcOperationalState.pause @1:1a40f361•e867⇵5d28✉0b4b2e6f (no payload)
info[20:16:42.405][Aspirapolvere Tavernetta]Pause (endpoint AspirapolvereTavernetta-1706100013688.1)
info[20:16:43.457][ProtocolService]Invoke » AspirapolvereTavernetta.AspirapolvereTavernetta-1706100013688.rvcOperationalState.pause @1:1a40f361•e867⇵5d28✉0b4b2e6f commandResponseState: { errorStateId: 0, errorStateDetails: "Fully operational" }
info[20:16:43.463][Transaction]Tx ◦setStateOf<AspirapolvereTavernetta.AspirapolvereTavernetta-1706100013688>#94 waiting on @1:1a40f361•e867⇵5d28✉0b4b2e6f
info[20:16:47.920][InteractionServer]Invoke « @1:1a40f361•e867⇵5d29 invokes: MA_roboticvacuumcleaner:0x1.RvcOperationalState:0x61.goHome:0x80
info[20:16:47.924][ProtocolService]Invoke « AspirapolvereTavernetta.AspirapolvereTavernetta-1706100013688.rvcOperationalState.goHome @1:1a40f361•e867⇵5d29✉0b4b2e72 (no payload)
info[20:16:47.927][Aspirapolvere Tavernetta]GoHome (endpoint AspirapolvereTavernetta-1706100013688.1)
info[20:16:47.930][Transaction]Tx ◦setStateOf<AspirapolvereTavernetta.AspirapolvereTavernetta-1706100013688>#95 waiting on @1:1a40f361•e867⇵5d29✉0b4b2e72

**Expected behavior**
the RVC to return to the dock

**Screenshots**
Please always add screenshots to help explain your problem. Possibly taken from the console output with colors.

**Debug file**
Please always add the debug file to help explain your problem.
[How to download the log](https://github.com/Luligu/matterbridge/blob/main/README.md#how-to-send-the-debug-log-files)

**Plugin (please complete the following information):**
This happens consistently with 3 different architectures, 3 different RVCs:

1) home assistant with Matterbridge 2026.5.3.
2) Raspberry PI 4- with following architecture:
Interface name	eth0
Mac address	dc:a6:32:ee:64:a7
IPv4 address	192.168.178.138
IPv6 address	fd00::5f31:63a9:ea00:4d9c
Node version	22.14.0
Hostname	raspberrypi
User	root
Os	Linux (5.10.92-v7l+)
Platform	linux (arm)

3) Raspberry PI 4 - with following architecture:
Interface name	eth0
Mac address	e4:5f:01:b5:6b:2c
IPv4 address	192.168.178.203
IPv6 address	fd00::e65f:1ff:feb5:6b2c
Node version	24.15.0
Hostname	matterbridge
User	root
Os	Linux (6.12.75+rpt-rpi-v8)
Platform	linux (arm64)

Problem has been reproduced on 3 different RVCs (all of them same model - sdjqr02rr - Xiaomi roborock mi robot - each one of them with its own individual Static IP

**Matterbridge (please complete the following information):**

- Mode [e.g. as a service or docker container]
- Version [e.g. 1.6.6]
- OS: [e.g. where you run Matterbridge]

**Additional context**

### PR DESCRIPTION
**Pull request by phuanh004**
## Description

This PR resolves the issue where pressing "Return to Dock" (or sending the `goHome` command) from Apple Home / Matter controllers hangs indefinitely and fails to return the robot to its dock (fixes #273).

### Root Cause
When a Matter controller invokes `goHome`, Matter.js executes the handler inside an active command transaction (`invoke`). In the previous implementation:
```typescript
this.endpoint.addCommandHandler(\x27goHome\x27, async () => {
  await this.endpoint?.updateAttribute(RvcOperationalState, \x27operationalState\x27, RvcOperationalState.OperationalState.SeekingCharger);
  await this.deviceManager.device.activateCharging();
});
```
Calling `await this.endpoint?.updateAttribute(...)` queues an attribute update transaction (`setStateOf`), which is blocked waiting for the parent `invoke` transaction to complete before acquiring the state lock. Because `updateAttribute` was awaited inside the handler, neither transaction could proceed, resulting in a circular transaction deadlock on Matterbridge:
```text
[InteractionServer] Invoke « ... rvcOperationalState.goHome
[ProtocolService] Invoke « RoborockS5...rvcOperationalState.goHome
[Transaction] Tx ◦setStateOf<RoborockS5...>#ce waiting on @1:...✉...
```
Matter controllers (such as Apple Home) eventually time out waiting for the command response.

### Changes
1. **Remove `await updateAttribute` inside `goHome`**: Rely on `stateChangedHandlers.in_returning` and `stateChangedHandlers.state` (which already handle updating `operationalState` to `SeekingCharger` reactively when the vacuum actually transitions to returning/docking). This eliminates the transaction deadlock completely.
2. **Handle `RvcRunMode.changeToMode(1)` (Idle mode)**: Fulfill the existing `TODO` in `RvcRunMode.changeToMode` by calling `activateCharging()` when Idle mode is requested, ensuring switching to Idle from Apple Home directs the vacuum to the dock.
3. **Prevent `in_returning` from overriding `paused` status**: When returning is paused, some firmwares continue reporting `in_returning: 1`. Checking `this.deviceManager.property(\x27state\x27) !== \x27paused\x27` ensures the `paused` operational state is preserved.
4. **Support `\x27disable\x27` timer status**: Accept `\x27disable\x27` (in addition to `\x27off\x27` and `\x27disabled\x27`) when discovering room segments from midnight timers on Roborock firmwares.
5. **Tests**: Updated and added unit tests in `src/vacuum_device_accessory.test.ts` covering Idle mode, deadlock-free `goHome`, paused returning state, and timer status parsing.

Special thanks to @nadorjozsef who previously investigated this in #273 and shared initial findings in their fork.

## Checklist

- [x] Update the CHANGELOG.md file, including a description of the changes, following the convention commented at the top of the file.
- [x] Add tests for the new code that you just changed. We'd like to keep the coverage as close to 100% as possible.
- [x] Add the appropriate labels to the PR.

### CODE REVIEW

File: src/vacuum_device_accessory.test.ts [hunk 1]
```diff
@@ -110,8 +110,8 @@ describe('VacuumDeviceAccessory', () => {
         deviceManagerMock.device.getTimer.mockResolvedValue([
           // This one will be discarded because it's ON
           ['timer-id', 'on', ['0 0 * * *', ['action', { segments: '16,17' }]]],
-          // This one is taken
-          ['timer-id', 'off', ['0 0 * * *', ['action', { segments: '16,17' }]]],
+          // This one is taken (supports off, disabled, disable)
+          ['timer-id', 'disable', ['0 0 * * *', ['action', { segments: '16,17' }]]],
```
Comment (afharo) [inline src/vacuum_device_accessory.test.ts:114]: nit: I'd prefer it if we had another test that covers the new state as well instead of changing the existing test.

**Review by afharo — APPROVED** [review]
Thank you for such a critical fix!

LGTM! 

Q: Do you know if there are other similar deadlock situations?

**Comment by phuanh004** [comment]
Thanks for the review and approval @afharo!

Regarding your question:
> **Q: Do you know if there are other similar deadlock situations?**

Yes! After auditing the codebase against Matter.js transaction semantics, here is a breakdown of why this happens and where else it applies:

---

### 1. The Core Deadlock Mechanism in Matter.js
In Matter.js, each command invocation (`InvokeRequest`) acquires an exclusive transaction lock on that cluster behavior (`lockOnInvoke: true` in `ProtocolService.ts`).

When an `addCommandHandler` callback calls `await endpoint.updateAttribute(TargetCluster, ...)` (or `setAttribute`):
1. `updateAttribute()` executes `endpoint.setStateOf(behavior, ...)`.
2. `setStateOf()` opens an internal transactional context (`this.act("setStateOf<...>")`) and tries to acquire exclusive resource access on that cluster behavior via `tx.addResources(behavior)`.
3. If `TargetCluster` is the **same cluster** currently executing the command, its resource is already locked by the parent `Invoke` transaction.
4. Matter.js suspends the state update transaction in a `waiting on` state (`tx.waitFor(...)` in `Tx.ts` / `ResourceSet.ts`).
5. Because the handler is `await`ing `updateAttribute()`, the command handler cannot finish until `updateAttribute()` resolves — and `updateAttribute()` cannot resolve until the command handler finishes and yields the lock.
6. **Result:** Circular transaction deadlock (`Tx setStateOf<...> waiting on @1:...`). The controller (e.g. Apple Home) eventually times out.

---

### 2. Audit of Other Command Handlers in This Plugin

#### 🟡 `ServiceArea.selectAreas` (Line 157–164) — Identical Pattern
```typescript
this.endpoint.addCommandHandler('selectAreas', async (data) => {
  this.log.debug(`Select areas command received: ${data.request.newAreas}`);
  let selectedAreas = data.request.newAreas;
  if ((data.attributes.supportedAreas as ServiceArea.Area[])?.length === selectedAreas.length) {
    selectedAreas = []; // Force empty if all areas are selected
  }
  await this.endpoint?.updateAttribute(ServiceArea, 'selectedAreas', selectedAreas);
});
```
* `selectAreas` is a command on the `ServiceArea` cluster, so `MatterbridgeServiceAreaServer` is locked during invocation.
* Calling `await this.endpoint?.updateAttribute(ServiceArea, 'selectedAreas', ...)` inside the handler attempts to lock the exact same cluster in a child `setStateOf` transaction.
* Furthermore, in `@matterbridge/core` (`MatterbridgeServiceAreaServer.selectAreas()`), it executes `return await super.selectAreas(request)`, which already natively updates `this.state.selectedAreas = request.newAreas` per Matter specification!
* **Recommendation:** The awaited `updateAttribute` inside `selectAreas` should be removed (or handled without an awaited `setStateOf`) to prevent the same transaction stall.

#### 🟢 `RvcRunMode.changeToMode` (Line 127) — Cross-Cluster Update
```typescript
await this.endpoint?.updateAttribute(ServiceArea, 'currentArea', selectedAreas[0]);
```
* Here, `currentArea` updates `ServiceArea` while the active command lock is on `RvcRunModeServer`. Because they are different clusters, it does not directly self-deadlock, though in general awaiting attribute changes inside command handlers is best avoided.

#### 🟢 Other Command Handlers (`stop`, `pause`, `resume`, `identify`, `RvcCleanMode.changeToMode`)
* **Completely safe.** None of them call `updateAttribute` inside their handlers. They dispatch the command to `deviceManager.device` (miIO) and rely on the reactive event stream (`this.deviceManager.stateChanged$`) in `postRegister()`, which runs asynchronously outside of any active Matter command transaction.

---

### 3. General Rule for Matterbridge Plugins
Inside `addCommandHandler(Command, async () => ...)`:
> **Never `await updateAttribute` (or `setAttribute`) for the same cluster server.** Let Matterbridge / Matter.js update its cluster state natively, or allow device state events (`stateChanged$`) to update attributes reactively after the command transaction settles.

---

*(P.S. I also pushed commit `5f5bed0` addressing Copilot's test suggestions — consolidating the duplicate Idle mode test and using `mockReturnValueOnce('paused')` so mock return values don't persist across test cases. 100% tests, linter, and build passing!)*
